### PR TITLE
feat(scaffolder): promote task recovery

### PR DESCRIPTION
**Pull request by dyatko**
## Summary

Revives and rebases [#32502](https://github.com/backstage/backstage/pull/32502) (originally by @benjdlambert) onto current `master` so task recovery can move out of its fragmented experimental configuration.

- adds the stable `scaffolder.taskRecovery` configuration
- preserves task secrets until a terminal state, allowing interrupted tasks to restart
- persists completed step outputs and resumes from the last known-good step
- keeps workspace persistence pluggable, with GCS support and a new development-only database provider
- retains backwards-compatible fallbacks for the existing `EXPERIMENTAL_*` configuration
- adds migration documentation, changesets, and recovery integration coverage

## Event-stream regression fixed

The previous draft derived `SerializedTaskEvent.isTaskRecoverable` from the global `scaffolder.taskRecovery.enabled` flag. This made every terminal completion look non-terminal to SSE clients, leaving completed task polling streams open indefinitely. Enough leaked streams can exhaust backend resources and prevent new task pages from receiving progress events.

This version keeps automatic crash recovery global, but derives event-stream recoverability from the task's legacy per-template `startOver` strategy. A new red/green regression test proves that enabling global recovery does not keep a normal task's completion stream alive.

## Why revive this

The existing `startOver` strategy still repeats all task steps after a worker crash. That is expensive for long-running templates and unsafe for actions whose external side effects are not fully idempotent. Persisting completed step outputs lets the worker skip known-good work and retry only the interrupted step, while checkpoints remain available for idempotency inside individual actions.

The earlier draft was closed after going stale rather than due to a rejected design. This PR preserves that design, resolves current-master conflicts, and incorporates the event-stream lifecycle issue discovered while operating the feature.

## Test plan

- [x] Event-stream regression test: red with global `isTaskRecoverable`, green with per-task derivation
- [x] `StorageTaskBroker.test.ts` — 27/27
- [x] `NunjucksWorkflowRunner.test.ts` — 90/90
- [x] recovery storage/workspace/helper tests — 44/44
- [x] database workspace provider — 7/7
- [x] GCP workspace provider — 4/4
- [x] router suite rerun in isolation — 37/37
- [x] `yarn lint --since HEAD`
- [x] `yarn tsc`
- [x] `yarn build:api-reports`
- [x] `yarn install --immutable`

The full changed-package test command also passed the recovery suites, but three unrelated router tests hit their 5-second timeout under parallel load before Jest/Node 20 ended with a stack overflow. The router suite passed 37/37 when rerun in isolation.

## Follow-up / review focus

This is intentionally a draft for Scaffolder SIG feedback, as requested on #32502. In particular, review is welcome on the stable config shape, workspace-provider packaging, and whether task recovery should remain globally enabled or support per-template strategy overrides.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: plugins/scaffolder-backend-module-workspace-database/src/module.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            'Database workspace provider is not supported in production and will not be enabled',
+          );
+          return;
+        }
+
+        const db = await database.getClient();
+
+        // Run migrations for our table
+        if (!database.migrations?.skip) {
+          await db.migrate.latest({
```
Comment (freben) [inline plugins/scaffolder-backend-module-workspace-database/src/module.ts:66]: A module cannot use the same migrations state table as the host plugin. They will clobber each other and run into trouble. See https://github.com/backstage/backstage/blob/8cc87f79a390bdc3ec4fa1341dda507e8ec646f4/plugins/catalog-backend-module-incremental-ingestion/src/database/migrations.ts#L29

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 2]
```diff
@@ -583,6 +592,13 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
         context.steps[step.id] as unknown as NunjitsuTemplateValue,
       );
 
+      // Persist step state for recovery
+      await task.updateStepState?.({
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Here we mark completed before the workspace gets serialized in the `finally`. Is there a risk that an error being thrown in between, leaves the state in a bad state without the workspace data?

File: plugins/scaffolder-node/src/actions/types.ts [hunk 3]
```diff
@@ -38,9 +38,12 @@ export type ActionContext<
...
   input: TActionInput;
-  checkpoint<T extends JsonValue | void>(
-    opts: CheckpointContext<T>,
-  ): Promise<T>;
+  /**
+   * Execute a function with checkpoint support for recovery.
+   * The return value is serialized to JSON for storage - non-serializable
+   * values (functions, circular refs, etc.) are safely stripped.
+   */
+  checkpoint<T>(opts: CheckpointContext<T>): Promise<T>;
```
Comment (freben) [inline plugins/scaffolder-node/src/actions/types.ts]: does this T need to be more constrained, such as extending `JsonValue`? since that's the real limitation when serializing

**Review by freben — CHANGES_REQUESTED** [review]

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts (hunk 2 above)
Comment (dyatko) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Fixed in 4178027972b6. Successful steps now serialize the workspace before `updateStepState` marks them completed. I also added tests that assert the ordering and prove a serialization failure never persists completed step state; failed/interrupted steps still serialize from `finally`.

File: plugins/scaffolder-node/src/actions/types.ts (hunk 3 above)
Comment (dyatko) [reply plugins/scaffolder-node/src/actions/types.ts]: Agreed, fixed in 4178027972b6. `CheckpointContext` and `ActionContext.checkpoint` now constrain `T extends JsonValue | void`, with the mock context and generated API reports updated as well. The safe runtime serialization remains defensive handling for JavaScript callers.

File: plugins/scaffolder-backend-module-workspace-database/src/module.ts (hunk 1 above)
Comment (dyatko) [reply plugins/scaffolder-backend-module-workspace-database/src/module.ts:66]: Fixed in 4178027972b6. The module now passes a dedicated `scaffolder_workspace_database_migrations` table name to Knex, so its migration state and lock tables cannot collide with the host scaffolder plugin.

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 4]
```diff
@@ -416,8 +406,6 @@ export class DatabaseTaskStore implements TaskStore {
         .update({
           status: 'processing',
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:430]: This now preserves secrets for every claimed task, even when `scaffolder.taskRecovery.enabled` is unset or false. Since recovery defaults off, ordinary tasks will retain tokens in the database until terminal completion (and indefinitely if a worker dies and no janitor completes the task), whereas they were previously cleared on claim. Could secret retention be gated on recovery being enabled so the default security lifecycle remains unchanged?

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 5]
```diff
@@ -670,10 +694,32 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
...
+          `Task recovered - resuming from last known good state. ${completedStepIds.length} step(s) already completed.`,
+        );
+      }
+
       let firstError: Error | undefined;
       const allErrors: Array<{ step: TaskStep; error: Error }> = [];
 
       for (const step of task.spec.steps) {
+        const savedStepState = savedSteps[step.id];
+        if (savedStepState?.status === 'completed') {
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:725]: Completed step state is consumed unconditionally here, while `TaskManager` also persists it unconditionally and `retryTask` does not clear `state`. This means a user retry of a failed task will skip previously completed steps even when task recovery is disabled, potentially without the workspace artifacts those steps produced. Could step-state persistence/resumption be gated by the recovery setting, or explicit retries clear this state?

**Review by freben — COMMENTED** [review]
A few remaining recovery-gating and event-consistency concerns:

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 6]
```diff
@@ -589,14 +601,26 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
         );
       }
 
+      workspaceSerializationAttempted = true;
+      await task.serializeWorkspace?.({ path: workspacePath });
+
+      // Persist step state for recovery
+      await task.updateStepState?.({
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: There is still a crash boundary between persisting the completed state here and `stepTrack.markSuccessful()` emitting the completed log event below. If the worker dies in that window, recovery skips the step but emits no replacement completion status. The frontend reconstructs step state from log events, so the original `processing` event remains the latest one and the recovered task UI shows this step as stuck processing. Could the recovery skip path emit a completed/recovered status event (or make state persistence and event emission atomic)?

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts (hunk 4 above)
Comment (dyatko) [reply plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:430]: Fixed in 1abb11cf7139. Secret retention on claim is now gated on recovery being enabled. `DatabaseTaskStore` clears secrets when moving a task to `processing` unless `scaffolder.taskRecovery.enabled` (with the `EXPERIMENTAL_recoverTasks` fallback) is set, or the task opted in via the legacy `startOver` strategy. The flag is resolved in the router via a shared `isTaskRecoveryEnabled` helper and passed into the store. So with recovery off (the default), the original security lifecycle is preserved and tokens are no longer retained past claim. Added tests covering both the disabled (cleared on claim) and enabled (retained until terminal) cases.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts (hunk 5 above)
Comment (dyatko) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:725]: Fixed in 1abb11cf7139. Both the persistence and the resumption of completed step state are now gated on recovery being enabled in `NunjucksWorkflowRunner`. When recovery is disabled the runner never calls `updateStepState` and never reads saved state, so a retry re-runs every step exactly as before, even if stale state happens to exist. When recovery is enabled it resumes from the last known-good step as intended. Added tests asserting no persistence and no resume-skip when recovery is disabled.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts (hunk 6 above)
Comment (dyatko) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Fixed in 1abb11cf7139. On resume, each skipped completed step now re-emits a `completed` status log event keyed by its `stepId`. Since the frontend reconstructs step status from log events, this fixes the recovered-step-stuck-in-`processing` UI and also closes the crash window you describe: the recovered run always re-emits the completion event for any step recorded as completed, so a crash between persisting the state and `stepTrack.markSuccessful()` is covered without needing cross-store atomicity. Added a regression test asserting the emitted event.

**Comment by freben** [comment]
For the schema type, it seems we run into trouble with mysql that ends up creating a tiny column limited to 64kB of data.  We'll need something like

```ts
if (knex.client.config.client.includes('mysql')) {
  table.specificType('workspace', 'LONGBLOB').notNullable();
} else {
  table.binary('workspace').notNullable();
}
```

**Comment by dyatko** [comment]
Addressed in 600f8b509de4. The workspace migration now uses `LONGBLOB` for MySQL clients and keeps the portable `binary` type for other databases, so MySQL can support the provider's documented workspace size instead of being limited to roughly 64 KiB.

I also added focused migration coverage for both the MySQL and non-MySQL branches. The workspace database module suite passes (9/9), along with lint and `yarn tsc`.

**Comment by dyatko** [comment]
Replaced by #35224, which uses the branch hosted directly in the backstage/backstage repository. All review fixes from this PR are included there.
